<a href="https://colab.research.google.com/github/brenda96011/IA-2026-2/blob/main/tema5/37_UNAM_AI_Actividad_Laboratorio_5_08_Predicci%C3%B3n_de_Texto_con_Transformers_y_BERT_entrenado.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<center>

$\Huge \textbf{Universidad Nacional Autónoma de México}$  
$\Huge \textbf{Facultad de Ciencias}$  
<p align="center">
  <img src="https://www.icat.unam.mx/wp-content/uploads/2021/11/Copia-de-LogoUNAM.-Azul.-Fondo-transparente.png" alt="UNAM" width="200"/>
</p>

<hr style="height:3px; background-color:#0B6E4F; border:none;"/>


$\LARGE \textbf{Inteligencia Artificial}$  

$\Large \textit{Laboratorio 5.8}$  


\begin{array}{rl}
\textbf{Docente:} & Dra. Jessica Sarahi Méndez Rincón \\[6pt]
\textbf{Ayudante de laboratorio:} & Diego Eduardo Peña Villegas \\[6pt]
\textbf{Alumna:} & Marisol Luna Méndez \\[6pt]
\textbf{Fecha de realización:} & 03/05/2026
\end{array}

</center>


**UNAM y la creadora del presente notebook, no guarda relación alguna con las marcas mencionadas como ejemplo. Las marcas son propiedad de sus titulares conforme a la legislación aplicable, se utilizan con fines académicos y didácticos, por lo que no existen fines de lucro, relación publicitaria o de patrocinio.

---

In [ ]:
!pip install transformers

In [ ]:
from transformers import pipeline, AutoTokenizer, AutoModelForSequenceClassification

# Cargar el modelo pre-entrenado y el tokenizador de BERT en español
#model_name = "dccuchile/bert-base-spanish-wwm-uncased"
# Este ya sabe qué es Positivo, Negativo y Neutral
# Muy robusto para español
model_name = "nlptown/bert-base-multilingual-uncased-sentiment"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSequenceClassification.from_pretrained(model_name)

# Crear un pipeline para la clasificación de texto
text_classification_pipeline = pipeline(
    'text-classification',
    model=model,
    tokenizer=tokenizer
)

# Ejecutar la clasificación de texto en una oración en español
#result = text_classification_pipeline('Esta película es muy buena, la recomiendo totalmente')
#print(result)

# Ejecutar la clasificación
result = text_classification_pipeline('Esta película es muy buena, la recomiendo totalmente')

# Acceder a los datos dentro de la lista y el diccionario
label = result[0]['label']
score = result[0]['score']

print(f"Resultado: {label}")
print(f"Confianza: {score:.4f}")

config.json:   0%|          | 0.00/953 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/39.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/669M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/201 [00:00<?, ?it/s]

Resultado: 5 stars
Confianza: 0.7252


1. El cambio en las etiquetas (Labels)
Al usar un modelo ajustado (probablemente el de nlptown o uno similar de reseñas), el sistema ya no usa códigos genéricos como LABEL_1.

"5 stars": El modelo ha clasificado la frase ("Esta película es muy buena, la recomiendo totalmente") con la puntuación máxima. Ha detectado un sentimiento altamente positivo.

2. La Confianza (Score)
0.7252 (72.5%): Es una métrica sólida. Indica que, de todas las opciones posibles (1, 2, 3, 4 o 5 estrellas), el modelo está bastante seguro de que pertenece a la categoría de 5.

In [ ]:
from transformers import AutoTokenizer, AutoModelForCausalLM

# Descargar el modelo pre-entrenado de GPT-2 en español
model_name = "mrm8488/spanish-gpt2"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

tokenizer_config.json:   0%|          | 0.00/226 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

added_tokens.json:   0%|          | 0.00/24.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/90.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/883 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/510M [00:00<?, ?B/s]

In [ ]:
#Cambiar este texto para 3 ejemplos
input_str = "Había una vez"

max_length = 100
# Generar texto con el modelo pre-entrenado de GPT-2 en español
input_ids = tokenizer.encode(input_str, return_tensors="pt")
sample_outputs = model.generate(
    input_ids,
    do_sample=True,
    max_length=max_length,
    top_k=50,
    top_p=0.95,
    temperature=1.0,
    num_return_sequences=1
)
generated_text = tokenizer.decode(sample_outputs[0], skip_special_tokens=True)



print(generated_text)

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


Había una vez en Nueva York, un padre con su hijo que se sentó al fondo de la mesa y dijo, "¡Oh, mira lo que has hecho!"¿Qué padre?El Sr. Fellner, ¿no?- Sí.- ¡Hola!- ¡Papá!¿Cómo estás?- ¿Cómo va el negocio?- Bien. ¿Cómo va eso?- Bueno, ya sabes cómo son las cosas.- ¡Oh, no!, ¡no!, ¡no!


In [ ]:
#Cambiar este texto para 3 ejemplos
input_str = "El hombre bicentenario es"

max_length = 100
# Generar texto con el modelo pre-entrenado de GPT-2 en español
input_ids = tokenizer.encode(input_str, return_tensors="pt")
sample_outputs = model.generate(
    input_ids,
    do_sample=True,
    max_length=max_length,
    top_k=50,
    top_p=0.95,
    temperature=1.0,
    num_return_sequences=1
)
generated_text = tokenizer.decode(sample_outputs[0], skip_special_tokens=True)



print(generated_text)

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


El hombre bicentenario es la historia del arte del Perú y la historia de las culturas en general.Se ha considerado como precursora del arte colonial e inició la evolución del arte español a partir del arte de los aborígenes.El período de las bellas artes de los siglos XVII y XVIII es rico en ejemplos de los diversos estilos pictóricos y de temática en concreto.En la segunda mitad del siglo XIX, el arte español del siglo XIX tiene características propias, como los motivos religiosos, la composición temática y


<hr/>
<footer style="text-align:center; font-size:12px; color:gray;">
© 2026 UNAM Facultado de Ciencias – Todos los derechos reservados

</footer>